docs/source/notes/autograd.rst

```diff
@@ -3,119 +3,394 @@
+as any other non-autograd library that provides inplace operations.
+To be able to compute the gradients, we consider that there are different versions of the tensor that existed and that the user tensor always points to the latest version.
+This allows us to write the automatic differentiation rules as before.
+
+See the two functions below for examples where we use subscripts to specify each version of the Tensor.:
+
+.. code::
+
+    def f(x):
+        z = op_1(x)
+        # z points to z_0
+        z.op_2_()
+        # z points to z_1
+        return z
+
+    def g(x):
+        z = op_1(x)
+        # z points to z_0
+        y = z.select(0, 0)
+        # y is a view into z
+        # y points to y_0 = z_0.select(0, 0)
+        z.op_2_()
+        # both z and y now points to y_1 = z_1.select(0, 0)
+        return y
+
+The corresponding autograd computations can simply be writen using the implicit variables: :math:`v^T J_f = v^T J_{op_2\_} J_{op_1} = v^T \dfrac{\partial{z_1}}{\partial{z_0}} \dfrac{\partial{z_0}}{\partial{x}} = v^T \dfrac{\partial{z_1}}{\partial{x}} = v^T \dfrac{\partial{z}}{\partial{x}}`.
+And for g, it is harder to write as only part of :code:`op_2_()` is used when computing :code:`y`, but it can be written as :math:`v^T J_g = v^T \dfrac{\partial{y_1}}{\partial{y_0}} \dfrac{\partial{y_0}}{\partial{z_0}} \dfrac{\partial{z_0}}{\partial{x}}`.
+
+Computing not gradients
+^^^^^^^^^^^^^^^^^^^^^^^
+
+In some cases, the user might need to compute something that is not the "true gradient" of its function.
+We identify few use cases here:
+
+- Ignore the gradients for some part of the computation as they could render the computed gradients unstable: gradients of square root near :math:`0`.
+- Ignore the gradients in a set of computations where we know they will be :math:`0` and so don't need to be computed: a set of operations that compute values that are the used as integer to index a tensor.
+- Update the value of a tensor that requires gradients without this update being recorded: initialization of the weights of a neural network.
+- Work with tensors that may or may not require gradients to perform operations that do not need gradients computed: optimizers for neural networks or method that compute the 0-1 accuracy for classification.
+
+To be able to perform all these operations, we provide three main constructs.
+
+Detaching
+"""""""""
+
+The first construct is the method :code:`y = x.detach()` that can be used on any tensor.
+This is a special function for which :math:`J_{detach} = \bm{0}`.
+This means that however :code:`y` is used, the contribution via :code:`y` to the gradient of :code:`x` is always going to be :math:`0`.
+
+.. code::
+
+    def f(x):
+        z = op_1(x)
+
+        y = z.detach()
+        return y
+
+    # x is a Tensor of the appropriate size
+    x.requires_grad_()
+
+    # v is a Tensor of the appropriate size
+    f(x).backward(v)
```

Comment:
You should add the outputs as comments here.

Reply:
Outputs?
Which information you expect to see?

Reply:
oops nvm :)